File: BOJ/28_트리/4803_트리.py
import sys
from collections import deque
read = sys.stdin.readline

def bfs(cur) :
    q = deque()
    q.append(cur)
    e,v = 0,0
    while q :
        here = q.popleft()
        if visited[here] == 1 : continue

        v += 1
        visited[here] = 1
        for n in adjs[here] :
            e += 1
            q.append(n)

    if e / 2 == v - 1 : return True
    else : return False
    
n,m = -1,-1
case = 0
while(1) :
    n,m = map(int,read().split())
    if n == 0 and m == 0 : break

    adjs = [[] for _ in range(n+1)]
    visited = [0 for _ in range(n+1)]
    for _ in range(m) :
        x,y = map(int, read().split())
        adjs[x].append(y)
        adjs[y].append(x)

    tree = 0
    for i in range(1,n+1) :
        if visited[i] == 1 : continue
        if bfs(i) : 
            tree += 1

    case += 1
    if tree == 0 :
        print("Case {}: No trees.".format(case))
    elif tree == 1 :
        print("Case {}: There is one tree.".format(case))
    else :
        print("Case {}: A forest of {} trees.".format(case,tree))

    

File: BOJ/28_트리/4803_트리.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <string.h>

#define MAXN 500

using namespace std;

vector<int> adjs[MAXN+1];
bool visited[MAXN+1];
int N,M,x,y;

bool bfs(int st) {

    queue<int> q;
    q.push(st);
    int v = 0, e = 0;
    while (!q.empty()) {
        int here = q.front();
        q.pop();

        if(visited[here]) continue;

        visited[here] = 1;
        v += 1;
        for(int i=0; i<adjs[here].size(); i++) {
            e += 1;
            q.push(adjs[here][i]);
        }
    }
    if(v-1 == e/2) return true;
    else return false;
}

int main() {
    int cnt = 0;
    while (1) {
        scanf("%d %d",&N,&M);
        if(N == 0 && M == 0) break;

        for(int i=0; i<=MAXN; i++) adjs[i].clear();
        memset(visited,false,sizeof(visited));

        for(int i=0; i<M; i++) {
            scanf("%d %d",&x,&y);
            adjs[x].push_back(y);
            adjs[y].push_back(x);
        }
        int tree = 0;
        for(int i=1; i<=N; i++) {
            if(visited[i]) continue;
            if(bfs(i)) tree+=1;
        }

        cnt += 1;
        if(tree == 0) printf("Case %d: No trees.\n",cnt);
        else if(tree == 1) printf("Case %d: There is one tree.\n",cnt);
        else printf("Case %d: A forest of %d trees.\n",cnt,tree);
    }
    return 0;
}